controllers/reclaimspacejob_controller.go

```diff
@@ -18,45 +18,332 @@ package controllers
+		logger.Info("Maximum retry limit reached")
+		rsJob.Status.Result = csiaddonsv1alpha1.OperationResultFailed
+		rsJob.Status.Message = "Maximum retry limit reached"
+		rsJob.Status.CompletionTime = &v1.Time{Time: time.Now()}
+	}
+
+	if statusErr := r.Client.Status().Update(ctx, rsJob); statusErr != nil {
+		logger.Error(statusErr, "failed to update status")
 
-	// TODO(user): your logic here
+		return ctrl.Result{}, statusErr
+	}
 
-	return ctrl.Result{}, nil
+	if rsJob.Status.Result != "" {
+		// since result is already set, just dequeue
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, err
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ReclaimSpaceJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&csiaddonsv1alpha1.ReclaimSpaceJob{}).
+		WithEventFilter(predicate.GenerationChangedPredicate{}).
 		Complete(r)
 }
+
+func (r *ReclaimSpaceJobReconciler) reconcile(
+	ctx context.Context,
+	logger *logr.Logger,
+	spec *csiaddonsv1alpha1.ReclaimSpaceJobSpec,
+	status *csiaddonsv1alpha1.ReclaimSpaceJobStatus,
+	namespace string) error {
+
+	if status.StartTime == nil {
+		status.StartTime = &v1.Time{Time: time.Now()}
+	} else {
+		// not first reconcile, increment retries
+		status.Retries++
+	}
+
+	if time.Now().After(status.StartTime.Time.Add(time.Second * time.Duration(spec.RetryDeadlineSeconds))) {
+		status.Result = csiaddonsv1alpha1.OperationResultFailed
+		status.Message = "Time limit reached"
+		status.CompletionTime = &v1.Time{Time: time.Now()}
+
+		return nil
+	}
+
+	driverName, pvName, nodeID, err := r.getTargetDetails(ctx, logger, spec.Target, namespace)
+	if err != nil {
+		logger.Error(err, "Failed to get target details")
+		status.Message = fmt.Sprintf("Failed to get target details: %v", err)
+		//setFailedStatus
+
+		return err
+	}
+
+	controllerFound, controllerReclaimedSpace, err := r.controllerReclaimSpace(ctx, logger, driverName, pvName)
+	if err != nil {
+		logger.Error(err, "Failed to make controller request")
+		//setFailedStatus
+
+		return err
+	}
+
+	nodeFound, nodeReclaimedSpace, err := r.nodeReclaimSpace(ctx, logger, driverName, nodeID, pvName)
+	if err != nil {
+		logger.Error(err, "Failed to make node request")
+		//setFailedStatus
+
+		return err
+	}
+
+	if !controllerFound && !nodeFound {
+		//setFailedStatus
+
+		return errors.New("Controller and Node Client not found")
+	}
+
+	reclaimedSpace := int64(0)
+	if controllerFound && controllerReclaimedSpace != nil {
+		reclaimedSpace = *controllerReclaimedSpace
+	}
+	if nodeFound && nodeReclaimedSpace != nil {
+		reclaimedSpace += *nodeReclaimedSpace
+	}
+
+	status.Result = csiaddonsv1alpha1.OperationResultSucceeded
+	status.Message = "Reclaim Space operation successfully completed."
+	status.ReclaimedSpace = *resource.NewQuantity(reclaimedSpace, resource.DecimalSI)
+
+	return nil
+}
+
+func (r *ReclaimSpaceJobReconciler) getTargetDetails(
```

Comment:
what are target details, and why return **THREE** strings? It would be clearer to return a struct `targetDetails`. And have documentation.